### ISSUE TITLE
Parse Failure when retriving operations by ledger

### ISSUE THREAD

**Issue by ivan71kmayshan27** [Synesso/scala-stellar-sdk#65]
When I'm trying to get all operations in specific ledger № 580147 of TestNetwork via `TestNetwork.operationsByLedger(580147)` I`m getting JSON parse exception(see it on testnet horizon https://horizon-testnet.stellar.org/ledgers/580147/operations?cursor=0&order=asc&limit=100)
```
Exception in thread "main" org.json4s.package$MappingException: unknown error
	at org.json4s.Extraction$.extract(Extraction.scala:44)
	at org.json4s.ExtractableJsonAstNode.extract(ExtractableJsonAstNode.scala:21)
	at stellar.sdk.inet.RawPage.$anonfun$parse$1(Page.scala:13)
	at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:237)
	at scala.collection.immutable.List.foreach(List.scala:392)
	at scala.collection.TraversableLike.map(TraversableLike.scala:237)
	at scala.collection.TraversableLike.map$(TraversableLike.scala:230)
	at scala.collection.immutable.List.map(List.scala:298)
	at stellar.sdk.inet.RawPage.parse(Page.scala:13)
	at stellar.sdk.inet.Horizon.$anonfun$getPage$2(HorizonAccess.scala:155)
	at scala.util.Success.$anonfun$map$1(Try.scala:255)
	at scala.util.Success.map(Try.scala:213)
	at scala.concurrent.Future.$anonfun$map$1(Future.scala:292)
	at scala.concurrent.impl.Promise.liftedTree1$1(Promise.scala:33)
	at scala.concurrent.impl.Promise.$anonfun$transform$1(Promise.scala:33)
	at scala.concurrent.impl.CallbackRunnable.run$$$capture(Promise.scala:64)
	at scala.concurrent.impl.CallbackRunnable.run(Promise.scala)
	at akka.dispatch.BatchingExecutor$AbstractBatch.processBatch(BatchingExecutor.scala:55)
	at akka.dispatch.BatchingExecutor$BlockableBatch.$anonfun$run$1(BatchingExecutor.scala:92)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
	at scala.concurrent.BlockContext$.withBlockContext(BlockContext.scala:85)
	at akka.dispatch.BatchingExecutor$BlockableBatch.run(BatchingExecutor.scala:92)
	at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:40)
	at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(ForkJoinExecutorConfigurator.scala:49)
	at akka.dispatch.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at akka.dispatch.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at akka.dispatch.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at akka.dispatch.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
Caused by: stellar.sdk.model.response.ResponseParseException: Unable to parse document:
{
  "_links":{
    "self":{
      "href":"https://horizon-testnet.stellar.org/operations/2491712391876609"
    },
    "transaction":{
      "href":"https://horizon-testnet.stellar.org/transactions/4565b9f5c356530c6c284b4f52dfa5ab8ba1d6d3aae783e066551b861c61ad1e"
    },
    "effects":{
      "href":"https://horizon-testnet.stellar.org/operations/2491712391876609/effects"
    },
    "succeeds":{
      "href":"https://horizon-testnet.stellar.org/effects?order=desc&cursor=2491712391876609"
    },
    "precedes":{
      "href":"https://horizon-testnet.stellar.org/effects?order=asc&cursor=2491712391876609"
    }
  },
  "id":"2491712391876609",
  "paging_token":"2491712391876609",
  "transaction_successful":true,
  "source_account":"GBSDPBNBWGQM4YZZKC3DVSO5H6ZB6EZK2H474VX5ILM5FIOSWS5HN3AX",
  "type":"set_options",
  "type_i":5,
  "created_at":"2019-04-02T21:00:35Z",
  "transaction_hash":"4565b9f5c356530c6c284b4f52dfa5ab8ba1d6d3aae783e066551b861c61ad1e",
  "signer_key":"TBAW357LCGCQ4AZ3S6LXZFB7QRGBNUSZZCRXQGFE3PSIO7BSYP7P6CO2",
  "signer_weight":2
}
	at stellar.sdk.model.response.ResponseParser$$anonfun$$lessinit$greater$1$$anonfun$apply$1.applyOrElse(ResponseParser.scala:13)
	at stellar.sdk.model.response.ResponseParser$$anonfun$$lessinit$greater$1$$anonfun$apply$1.applyOrElse(ResponseParser.scala:8)
	at scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:38)
	at org.json4s.CustomSerializer$$anonfun$deserialize$2.applyOrElse(Formats.scala:449)
	at org.json4s.CustomSerializer$$anonfun$deserialize$2.applyOrElse(Formats.scala:447)
	at org.json4s.Extraction$.customOrElse(Extraction.scala:658)
	at org.json4s.Extraction$.extract(Extraction.scala:402)
	at org.json4s.Extraction$.extract(Extraction.scala:40)
	... 27 more
Caused by: org.json4s.package$MappingException: unknown error
	at org.json4s.Extraction$.extract(Extraction.scala:44)
	at org.json4s.ExtractableJsonAstNode.extract(ExtractableJsonAstNode.scala:21)
	at stellar.sdk.model.op.TransactedOperationDeserializer$$anonfun$$lessinit$greater$1.apply(Transacted.scala:26)
	at stellar.sdk.model.op.TransactedOperationDeserializer$$anonfun$$lessinit$greater$1.apply(Transacted.scala:17)
	at stellar.sdk.model.response.ResponseParser$$anonfun$$lessinit$greater$1$$anonfun$apply$1.applyOrElse(ResponseParser.scala:11)
	... 34 more
Caused by: stellar.sdk.model.response.ResponseParseException: Unable to parse document:
{
  "_links":{
    "self":{
      "href":"https://horizon-testnet.stellar.org/operations/2491712391876609"
    },
    "transaction":{
      "href":"https://horizon-testnet.stellar.org/transactions/4565b9f5c356530c6c284b4f52dfa5ab8ba1d6d3aae783e066551b861c61ad1e"
    },
    "effects":{
      "href":"https://horizon-testnet.stellar.org/operations/2491712391876609/effects"
    },
    "succeeds":{
      "href":"https://horizon-testnet.stellar.org/effects?order=desc&cursor=2491712391876609"
    },
    "precedes":{
      "href":"https://horizon-testnet.stellar.org/effects?order=asc&cursor=2491712391876609"
    }
  },
  "id":"2491712391876609",
  "paging_token":"2491712391876609",
  "transaction_successful":true,
  "source_account":"GBSDPBNBWGQM4YZZKC3DVSO5H6ZB6EZK2H474VX5ILM5FIOSWS5HN3AX",
  "type":"set_options",
  "type_i":5,
  "created_at":"2019-04-02T21:00:35Z",
  "transaction_hash":"4565b9f5c356530c6c284b4f52dfa5ab8ba1d6d3aae783e066551b861c61ad1e",
  "signer_key":"TBAW357LCGCQ4AZ3S6LXZFB7QRGBNUSZZCRXQGFE3PSIO7BSYP7P6CO2",
  "signer_weight":2
}
	at stellar.sdk.model.response.ResponseParser$$anonfun$$lessinit$greater$1$$anonfun$apply$1.applyOrElse(ResponseParser.scala:13)
	at stellar.sdk.model.response.ResponseParser$$anonfun$$lessinit$greater$1$$anonfun$apply$1.applyOrElse(ResponseParser.scala:8)
	at scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:38)
	at org.json4s.CustomSerializer$$anonfun$deserialize$2.applyOrElse(Formats.scala:449)
	at org.json4s.CustomSerializer$$anonfun$deserialize$2.applyOrElse(Formats.scala:447)
	at org.json4s.Extraction$.customOrElse(Extraction.scala:658)
	at org.json4s.Extraction$.extract(Extraction.scala:402)
	at org.json4s.Extraction$.extract(Extraction.scala:40)
	... 38 more
Caused by: java.lang.IllegalArgumentException: requirement failed
	at scala.Predef$.require(Predef.scala:268)
	at stellar.sdk.model.PreAuthTxnSigner.<init>(Signer.scala:30)
	at stellar.sdk.model.op.OperationDeserializer$$anonfun$$lessinit$greater$1.$anonfun$apply$2(Operation.scala:143)
	at scala.util.Failure.getOrElse(Try.scala:222)
	at stellar.sdk.model.op.OperationDeserializer$$anonfun$$lessinit$greater$1.$anonfun$new$10(Operation.scala:143)
	at stellar.sdk.model.op.OperationDeserializer$$anonfun$$lessinit$greater$1.$anonfun$new$10$adapted(Operation.scala:141)
	at scala.Option.map(Option.scala:163)
	at stellar.sdk.model.op.OperationDeserializer$$anonfun$$lessinit$greater$1.$anonfun$new$9(Operation.scala:141)
	at scala.Option.flatMap(Option.scala:188)
	at stellar.sdk.model.op.OperationDeserializer$$anonfun$$lessinit$greater$1.apply(Operation.scala:140)
	at stellar.sdk.model.op.OperationDeserializer$$anonfun$$lessinit$greater$1.apply(Operation.scala:55)
	at stellar.sdk.model.response.ResponseParser$$anonfun$$lessinit$greater$1$$anonfun$apply$1.applyOrElse(ResponseParser.scala:11)
	... 45 more
```
Does testnet contains corrupted data or the client can`t parse something?

**Comment by Synesso** [comment]
The problem here is that the SDK is assuming every signer is an account id. This set options operation is setting a pre-auth-transaction as a signer. As the response object doesn't identify the type of signer, I could change this type to String and leave it up to the user to figure out what it is. I'd rather type it correctly, so I'm looking into that possibility. But chances are this will become a String. The user will need to then look up the source account to see which signer that String value matches.

**Comment by Synesso** [comment]
I think Horizon should provide a discriminator on this response. Until it does, we'll have to improvise. I've made a change that attempts to resolve the signer as a public key. If that fails, it assumes it's a pre-auth txn. This approach is incorrect, but will at least prevent the document from failing to parse.

**Comment by Synesso** [comment]
@ivan71kmayshan27 Thank you for raising this bug. I appreciate it. Please try v0.6.1 and let me know if it fixes it for you.

**Comment by ivan71kmayshan27** [comment]
@Synesso , unfortunately it fails again on 0.6.1 with the same exception. I've updated stack trace because I noticed that it was not full enough. From the line: 
```at stellar.sdk.model.PreAuthTxnSigner.<init>(Signer.scala:30)```
I can guess that it fails in `stellar.sdk.model.PreAuthTxnSigner` in line `30` in assert `require (bytes.length == 32)`, because `signer_key` of this (json in exception) document is 42 bytes instead of 32. In 0.6.1  `stellar.sdk.model.Signer.scala`remained unchanged.

**Comment by Synesso** [comment]
Sorry. My generated test data wasn't diverse enough to catch this. I should have tested with the data you provided.

I'm currently trying to figure out why these values are 42 bytes when the Java SDK specifies (on creation) that they be exactly 32 bytes.

**Comment by Synesso** [comment]
Tested fixed in v0.6.2.
